poppy/wfe.py

```diff
@@ -353,6 +353,151 @@ def get_opd(self, wave):
+        cen = int(self.screen_size/2)
+        maskY, maskX = np.mgrid[-cen:cen, -cen:cen]
+        ky = maskY*dk.value
+        kx = maskX*dk.value
+        k_map = np.sqrt(kx**2 + ky**2) # unitless for the math, but actually 1/m
+        
+        # calculate the PSD
+        psd = np.zeros_like(k_map) # initialize the total PSD matrix
+        for n in range(0, len(self.psd_weight)):
+            # loop-internal localized PSD variables
+            alpha = self.psd_parameters[n][0]
+            beta = self.psd_parameters[n][1]
+            outer_scale = self.psd_parameters[n][2]
+            inner_scale = self.psd_parameters[n][3]
+            surf_roughness = self.psd_parameters[n][4]
+            
+            # unit check
+            psd_units = beta.unit / ((dk.unit**2)**(alpha/2))
+            assert surf_roughness.unit == psd_units, "PSD parameter units are not consistent, please re-evaluate parameters."
+            surf_unit = (psd_units*(dk.unit**2))**(0.5)
+            
+            # initialize loop-internal PSD matrix
+            psd_local = np.zeros_like(psd)
+            
+            # Calculate the PSD equation denominator based on outer_scale presence
+            if outer_scale.value == 0: # skip out or else PSD explodes
+                # temporary overwrite of k_map at k=0 to stop div/0 problem
+                k_map[cen][cen] = 1*dk.value
+                # calculate PSD as normal
+                psd_denom = (k_map**2)**(alpha/2)
+                # overwrite PSD at k=0 to be 0 instead of the original infinity
+                psd_denom[cen][cen] = 0*psd_denom #.unit #no units in any calculation!
+                # return k_map to original state
+                k_map[cen][cen] = 0
+            else:
+                psd_denom = ((outer_scale.value**(-2)) + (k_map**2))**(alpha/2) # unitless currently
+            
+            psd_local = (beta.value*np.exp(-((k_map*inner_scale)**2))/psd_denom) + surf_roughness.value
+            
+            # apply as the sum with the weight of the PSD model
+            psd = psd + (self.psd_weight[n] * psd_local) # this should all be m2 m2, but stay unitless for all calculations
+        
+        # set the random noise
+        psd_random = np.random.RandomState()
+        psd_random.seed(self.seed)
+        rndm_noise = np.fft.fftshift(np.fft.fft2(psd_random.normal(size=(self.screen_size, self.screen_size))))
+        
+        psd_scaled = (np.sqrt(psd/(pixelscale.value**2)) * rndm_noise)
+        opd = ((np.fft.ifft2(np.fft.ifftshift(psd_scaled)).real*surf_unit).to(u.m)).value 
+        
+        if self.apply_reflection == True:
+            opd = opd/2
+            
+        if self.screen_size > wave.shape[0]: # crop to wave shape
+            opd = utils.pad_or_crop_to_shape(array=opd, target_shape=wave.shape)
+            
+        if self.wfe is not None:
+            rms = np.sqrt(np.mean(np.square(opd)))
+            opd = opd * (self.wfe.to(u.m).value/rms)
```

Comment:
Minor, but this can be rewritten as `opd *= self.wfe.to(u.m).value/rms`, which is equivalent but slightly more performant.

Reply:
Oh no you're seeing my rudimentary python coding. I've fixed this through my code. Thanks for the tip.

Reply:
No worries! We're all learning as we go. :-)

(I have spent a probably-disproportionate amount of time pursuing small gains in runtime throughout poppy, for lots of years now, so all these little optimizations are likely more apparent to me than they would be to most people)